core/opendaq/signal/include/opendaq/signal_impl.h

```diff
@@ -262,6 +311,26 @@ ErrCode SignalBase<TInterface, Interfaces...>::setDescriptor(IDataDescriptor* de
                 if (signalPtr.assigned())
                     valueSignalsOfDomainSignal.push_back(std::move(signalPtr));
             }
+
+            if (dataDescriptor.getSampleType() == SampleType::Struct)
+            {
+                auto typeManager = context.getTypeManager();
+                const auto name = dataDescriptor.getName();
+                const StructTypePtr structType = createTypeFromDescriptor(dataDescriptor);
+
+                if (!typeManager.hasType(name))
```

Comment:
I see an issue here with multithreaded access. But it more or less relates to making type manager threadsafe, which IMHO should have atomic function like `typeManager.tryAddType` and it should return either "Success", "ExistsSame", "ExistsDifferent" status codes.

Reply:
As this problem is out of scope for the given commit, I've added an [appropriate issue on Jira](https://blueberrydaq.atlassian.net/browse/TBBAS-1454).